File: LP-1/Assembler (Python)/Pass-1.py
# using https://github.com/Prago2001/LP-1/blob/main/SPOS/A1/pass_1.py

from os import sep, write

tab = "\t"
newline = "\n"

class Mnemonics:
 
    # constructor with all the opcodes
    def __init__(self):
        self.AD = {"START": 1,
                    "END": 2,
                    "ORIGIN": 3,
                    "EQU": 4,
                    "LTORG": 5}

        self.IS = {"STOP": 0,
                    "ADD": 1,
                    "SUB": 2,
                    "MULT": 3,
                    "MOVER": 4,
                    "MOVEM": 5,
                    "COMP": 6,
                    "BC": 7,
                    "DIV": 8,
                    "READ": 9,
                    "PRINT": 10}

        self.DL = {"DC": 1,
                   "DS": 2}

        self.REG = {"AREG": 1,
                    "BREG": 2,
                    "CREG": 3,
                    "DREG": 4}

        self.CC = {"LT": 1,
                    "LE": 2,
                    "EQ": 3,
                    "GT": 4,
                    "GE": 5,
                    "ANY": 6}


    def getClassType(self, string):
        if string in self.AD:
            return "AD"
        elif string in self.IS:
            return "IS"
        elif string in self.DL:
            return "DL"
        elif string in self.REG:
            return "RG"
        elif string in self.CC:
            return "CC"
        else:
            return ""


    def getCode(self, string):
        if string in self.AD:
            return self.AD[string]
        elif string in self.IS:
            return self.IS[string]
        elif string in self.DL:
            return self.DL[string]
        elif string in self.REG:
            return self.REG[string]
        elif string in self.CC:
            return self.CC[string]
        else:
            return -1



class Pass1:
    def __init__(self):
        self.opTab = Mnemonics()  # object of Mnemonics class for lookup
        self.symTab = {}
        self.litTab = {}
        self.poolTab = [0]  # default entry for 1st literal pool
        self.litIndex = 0  # literal table index (for appending)
        self.litPtr = 0  # literal table pointer (for tracking processed literals)
        self.LC = 0  # location counter
        self.IC = []

        # files
        self.inputFile = open("input.txt", "r")
        self.litTabFile = open("literalTable.txt","w")
        self.symTabFile = open("symbolTable.txt","w")
        self.poolTabFile = open("poolTable.txt","w")
        self.ICFile = open("intermediateCode.txt","w")


    # calculates and returns location for ORIGIN 
    def location(self, string):
        if "+" in string:
            string = string.split("+")  #  eg. [LOOP, 2] for LOOP+2
            return self.symTab[string[0]] + int(string[1])
        elif "-" in string:
            string = string.split("-")
            return self.symTab[string[0]] - int(string[1])
        else:
            return self.symTab[string]

    
    def compute(self):
        for line in self.inputFile.readlines():

            self.IC.append([])

            line = line.strip("\n")
            line = line.split("\t")

            label = line[0]
            opcode = line[1]
            if len(line) >= 3:
                operand1 = line[2]
            if len(line) >= 4:
                operand2 = line[3]


            # for labels
            if label:
                self.symTab[label] = self.LC  # set label, location


            # for opcodes that need specific handling
            if opcode == "START":
                self.LC = int(operand1)
                self.IC[-1].append(('AD', 1))
                self.IC[-1].append(('C', int(operand1)))
            
            elif opcode == "LTORG":
                for i in range(self.poolTab[-1], len(self.litTab)):
                    self.litTab[i][1] = self.LC  # set the location for literal 
                    self.IC[-1].append(('DL', 1))
                    self.IC[-1].append(('C', self.litTab[i][0]))
                    self.IC[-1].append(self.LC)
                    self.LC += 1
                    self.litPtr += 1

                    if i < (len(self.litTab) - 1):
                        self.IC.append([])
                
                self.poolTab.append(self.litPtr)

            elif opcode == "ORIGIN":
                self.LC = self.location(operand1)
                self.IC[-1].append(('AD', 3))
                self.IC[-1].append(('C', self.LC))
            
            elif opcode == "EQU":
                newLocation = self.location(operand1)
                self.symTab[label] = newLocation
                self.IC[-1].append(('AD', 4))
                self.IC[-1].append(('C', newLocation))

            elif opcode == "DC":
                self.IC[-1].append(('DL', 1))
                self.IC[-1].append(('C', int(operand1)))
                self.IC[-1].append(self.LC)
                self.LC += 1

            elif opcode == "DS":
                self.IC[-1].append(('DL', 2))
                self.IC[-1].append(('C', int(operand1)))
                self.IC[-1].append(self.LC)
                self.LC += int(operand1)

            elif opcode == "STOP":
                self.IC[-1].append(('IS', 0))
                self.IC[-1].append(self.LC)
                self.LC += 1

            elif opcode == "END":
                self.IC[-1].append(('AD', 2))
                # process remaining literals if any
                if self.litPtr != len(self.litTab):
                    for i in range(self.poolTab[-1], len(self.litTab)):
                        self.IC.append([])
                        self.litTab[i][1] = self.LC
                        self.IC[-1].append(('DL', 1))
                        self.IC[-1].append(('C', self.litTab[i][0]))
                        self.IC[-1].append(self.LC)
                        self.LC += 1
                        self.litPtr += 1
                    self.poolTab.append(self.litPtr)

            elif opcode == "BC":
                self.IC[-1].append(('IS', 7))
                # get class and code for 1st operand
                classType = self.opTab.getClassType(operand1)
                machineCode = self.opTab.getCode(operand1)
                self.IC[-1].append((classType, machineCode))
                # compute 2nd operand if exists
                if operand2 not in self.symTab:
                    self.symTab[operand2] = None
                symTabKeys = list(self.symTab.keys())
                self.IC[-1].append(("S",symTabKeys.index(operand2)))
                self.IC[-1].append(self.LC)
                self.LC += 1
                
            elif opcode == "READ":
                self.IC[-1].append(('IS', 9))
                self.symTab[operand1] = None
                symTabKeys = list(self.symTab.keys())
                self.IC[-1].append(("S",symTabKeys.index(operand1)))
                self.IC[-1].append(self.LC)
                self.LC += 1
            
            elif opcode == "PRINT":
                self.IC[-1].append(('IS', 10))
                symTabKeys = list(self.symTab.keys())
                self.IC[-1].append(('S', symTabKeys.index(operand1)))
                self.IC[-1].append(self.LC)
                self.LC += 1
            
            else:
                # for opcodes
                classType = self.opTab.getClassType(opcode)
                machineCode = self.opTab.getCode(opcode)
                self.IC[-1].append((classType, machineCode))

                # for operand1
                classType = self.opTab.getClassType(operand1)
                machineCode = self.opTab.getCode(operand1)
                self.IC[-1].append((classType, machineCode))

                # for operand2
                if "=" in operand2:  # if it is a literal
                    constant = operand2.strip("=")
                    constant = int(constant.strip("'"))
                    self.litTab[self.litIndex] = [constant, None]
                    self.IC[-1].append(('L', self.litIndex))
                    self.IC[-1].append(self.LC)
                    self.litIndex += 1
                else:
                    if operand2 in self.symTab:
                        symTabKeys = list(self.symTab.keys())
                        self.IC[-1].append(('S', symTabKeys.index(operand2)))
                        self.IC[-1].append(self.LC)
                    else:
                        self.symTab[operand2] = None
                        symTabKeys = list(self.symTab.keys())
                        self.IC[-1].append(('S', symTabKeys.index(operand2)))
                        self.IC[-1].append(self.LC)
                
                self.LC += 1
        
        # print all tables
        self.printST()
        self.printLT()
        self.printPT()
        self.printIC()

    def printST(self):
        print("\n--------------------")
        print("SYMBOL TABLE")
        print("--------------------")
        for index, item in enumerate(self.symTab):
            line = str(index) + tab + str(item) + tab + str(self.symTab[item]) + newline
            print(line, end="")
            self.symTabFile.write(line)
        self.symTabFile.close()
        print()


    def printLT(self):
        print("\n--------------------")
        print("LITERAL TABLE")
        print("--------------------")
        for item in range(len(self.litTab)):
            line = str(item) + tab + str(self.litTab[item][0]) + tab + str(self.litTab[item][1]) + newline
            print(line, end="")
            self.litTabFile.write(line)
        self.litTabFile.close()
        print()
 

    def printPT(self):
        print("\n--------------------")
        print("POOL TABLE")
        print("--------------------")
        for item in range(len(self.poolTab)):
            print(self.poolTab[item])
            self.poolTabFile.write(str(self.poolTab[item]) + newline)
        self.poolTabFile.close()
        print()


    def printIC(self):
        print("\n----------------------------------------------------")
        print("INTERMEDIATE CODE")
        print("----------------------------------------------------")
        for item in self.IC:
            line = ""
            for i in range(len(item)):
                line += str(item[i])
                if i!= len(item):
                    line += tab
            line += newline
            print(line, end="")
            self.ICFile.write(line)
        self.ICFile.close()       


obj = Pass1()
obj.compute()

File: LP-1/Assembler (Java)/Pass1.java
import java.io.*;
import java.util.*;

class Mnemonics
{
    private Map<String, Integer> AD = new HashMap<>();
    private Map<String, Integer> IS = new HashMap<>();
    private Map<String, Integer> DL = new HashMap<>();
    private Map<String, Integer> RG = new HashMap<>();
    private Map<String, Integer> CC = new HashMap<>();

    // add all mnemonics in constructor
    public Mnemonics() {
        AD.put("START", 1);
        AD.put("END", 2);
        AD.put("ORIGIN", 3);
        AD.put("EQU", 4);
        AD.put("LTORG", 5);

        IS.put("STOP", 0);
        IS.put("ADD", 1);
        IS.put("SUB", 2);
        IS.put("MULT", 3);
        IS.put("MOVER", 4);
        IS.put("MOVEM", 5);
        IS.put("COMP", 6);
        IS.put("BC", 7);
        IS.put("DIV", 8);
        IS.put("READ", 9);
        IS.put("PRINT", 10);

        DL.put("DC", 1);
        DL.put("DS", 2);

        RG.put("AREG", 1);
        RG.put("BREG", 2);
        RG.put("CREG", 2);
        RG.put("DREG", 2);

        CC.put("LT", 1);
        CC.put("LE", 2);
        CC.put("EQ", 3);
        CC.put("GT", 4);
        CC.put("GE", 5);
        CC.put("ANY", 6);
    }

    public String getClassType(String string)
    {
        if (AD.containsKey(string)) {
            return "AD";
        }
        else if (IS.containsKey(string)) {
            return "IS";
        }
        else if (DL.containsKey(string)) {
            return "DL";
        }
        else if (RG.containsKey(string)) {
            return "RG";
        }
        else if (CC.containsKey(string)) {
            return "CC";
        }
        else {
            return "";
        }
    }

    public int getMachineCode(String string)
    {
        if (AD.containsKey(string)) {
            return AD.get(string);
        }
        else if (IS.containsKey(string)) {
            return IS.get(string);
        }
        else if (DL.containsKey(string)) {
            return DL.get(string);
        }
        else if (RG.containsKey(string)) {
            return RG.get(string);
        }
        else if (CC.containsKey(string)) {
            return CC.get(string);
        }
        else {
            return -1;
        }
    }
}


class IntermediateCodeEntry {
    String type;
    Object value;

    IntermediateCodeEntry(String type, Object value) {
        this.type = type;
        this.value = value;
    }
}



public class Pass1
{
    private Mnemonics opTab = new Mnemonics();
    private Map<String, Integer> symTab = new HashMap<>();
    private List<int[]> litTab = new ArrayList<>();
    private List<Integer> poolTab = new ArrayList<>(Collections.singletonList(0));
    private int litIndex = 0;
    private int litPtr = 0;
    private int LC = 0;
    private List<List<IntermediateCodeEntry>> IC;

    private FileWriter ICFile;  
    private FileWriter symTabFile;
    private FileWriter litTabFile;
    private FileWriter poolTabFile;
    private BufferedReader inputFile;

    // open all files in constructor
    public Pass1() throws IOException {
        ICFile = new FileWriter("intermediateCode.txt");
        symTabFile = new FileWriter("symbolTable.txt");
        litTabFile = new FileWriter("literalTable.txt");
        poolTabFile = new FileWriter("poolTable.txt");
        inputFile = new BufferedReader (new FileReader("test-case-1.txt"));
        IC = new ArrayList<>();
    }


    public int location(String string) {
        if (string.contains("+")) {
            String[] parts = string.split("\\+");
            return symTab.get(parts[0]) + Integer.parseInt(parts[1]);
        }
        else if (string.contains("-")) {
            String[] parts = string.split("-");
            return symTab.get(parts[0]) - Integer.parseInt(parts[1]);
        }
        else {
            return symTab.get(string);
        }
    }


    public void compute() throws IOException {
        String line;
        while ((line = inputFile.readLine()) != null) {
            IC.add(new ArrayList<>());
    
            line = line.trim();
            String[] parts = line.split("\t");
    
            String label = parts[0];
            String opcode = parts[1];
            String operand1 = parts.length >= 3 ? parts[2] : null;
            String operand2 = parts.length >= 4 ? parts[3] : null;
            

            // for labels
            if (!label.isEmpty()) {
                symTab.put(label, LC);
            }


            // for opcodes that need to be handled separately
            if (opcode.equals("START")) {
                LC = Integer.parseInt(operand1);
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("AD", 1));
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("C", LC));
            }
            else if (opcode.equals("LTORG")) {
                for (int i = poolTab.get(poolTab.size() - 1); i < litTab.size(); i++) {
                    litTab.get(i)[1] = LC;
                    IC.get(IC.size() - 1).add(new IntermediateCodeEntry("DL", 1));
                    IC.get(IC.size() - 1).add(new IntermediateCodeEntry("C", litTab.get(i)[0]));
                    IC.get(IC.size() - 1).add(new IntermediateCodeEntry("LC", LC));
                    LC++;
                    litPtr++;
                    if (i < litTab.size() - 1) {
                        IC.add(new ArrayList<>());
                    }
                }
                poolTab.add(litPtr);
            }
            else if (opcode.equals("ORIGIN")) {
                LC = location(operand1);
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("AD", 3));
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("C", LC));
            }
            else if (opcode.equals("EQU")) {
                int newLocation = location(operand1);
                symTab.put(label, newLocation);
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("AD", 4));
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("C", newLocation));
            }
            else if (opcode.equals("DC")) {
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("DL", 1));
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("C", Integer.parseInt(operand1)));
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("LC", LC));
                LC++;
            }
            else if (opcode.equals("DS")) {
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("DL", 2));
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("C", Integer.parseInt(operand1)));
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("LC", LC));
                LC += Integer.parseInt(operand1);
            }
            else if (opcode.equals("STOP")) {
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("IS", 0));
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("LC", LC));
                LC++;
            }
            else if (opcode.equals("END")) {
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("AD", 2));
                if (litPtr != litTab.size()) {
                    for (int i = poolTab.get(poolTab.size() - 1); i < litTab.size(); i++) {
                        IC.add(new ArrayList<>());
                        litTab.get(i)[1] = LC;
                        IC.get(IC.size() - 1).add(new IntermediateCodeEntry("DL", 1));
                        IC.get(IC.size() - 1).add(new IntermediateCodeEntry("C", litTab.get(i)[0]));
                        IC.get(IC.size() - 1).add(new IntermediateCodeEntry("LC", LC));
                        LC++;
                        litPtr++;
                    }
                    poolTab.add(litPtr);
                }
            }
            else if (opcode.equals("BC")) {
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("IS", 7));
                String classType = opTab.getClassType(operand1);
                int machineCode = opTab.getMachineCode(operand1);
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry(classType, machineCode));
                if (!symTab.containsKey(operand2)) {
                    symTab.put(operand2, null);
                }
                List<String> symTabKeys = new ArrayList<>(symTab.keySet());
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("S", symTabKeys.indexOf(operand2)));
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("LC", LC));
                LC++;
            }
            else if (opcode.equals("READ")) {
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("IS", 9));
                symTab.put(operand1, null);
                List<String> symTabKeys = new ArrayList<>(symTab.keySet());
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("S", symTabKeys.indexOf(operand1)));
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("LC", LC));
                LC++;
            }
            else if (opcode.equals("PRINT")) {
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("IS", 10));
                List<String> symTabKeys = new ArrayList<>(symTab.keySet());
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("S", symTabKeys.indexOf(operand1)));
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry("LC", LC));
                LC++;
            }
            else {
                String classType = opTab.getClassType(opcode);
                int machineCode = opTab.getMachineCode(opcode);
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry(classType, machineCode));

                classType = opTab.getClassType(operand1);
                machineCode = opTab.getMachineCode(operand1);
                IC.get(IC.size() - 1).add(new IntermediateCodeEntry(classType, machineCode));
                
                try {
                    if (operand2.contains("=")){
                        int constant = Integer.parseInt(operand2.replace("=", "").replace("'", ""));
                        litTab.get(litIndex)[0] = constant;
                        IC.get(IC.size() - 1).add(new IntermediateCodeEntry("L", litIndex));
                        IC.get(IC.size() - 1).add(new IntermediateCodeEntry("LC", LC));
                        litIndex++;
                    }
                    else {
                        if (!symTab.containsKey(operand2)) {
                            symTab.put(operand2, null);
                        }
                        List<String> symTabKeys = new ArrayList<>(symTab.keySet());
                        IC.get(IC.size() - 1).add(new IntermediateCodeEntry("S", symTabKeys.indexOf(operand2)));
                        IC.get(IC.size() - 1).add(new IntermediateCodeEntry("LC", LC));
                    }
                }
                catch (NullPointerException n) {
                    n.getStackTrace();
                }
                
                LC++;
            }
        }

        // print all tables
        printST();
        printLT();
        printPT();
        printIC();
    }


    public void printST() {
        System.out.println("\n--------------------");
        System.out.println("SYMBOL TABLE");
        System.out.println("--------------------");

        try {
            for (Map.Entry<String, Integer> entry: symTab.entrySet()) {
                String line = entry.getKey() + "\t" + entry.getValue() + "\n";
                System.out.print(line);
                symTabFile.write(line);
            }
            symTabFile.close();
        }
        catch (IOException e) {
            e.printStackTrace();
        }
        System.out.println();
    }


    public void printLT() {
        System.out.println("\n--------------------");
        System.out.println("LITERAL TABLE");
        System.out.println("--------------------");

        try {
            for (int item = 0; item < litTab.size(); item++) {
                String line = item + "\t" + litTab.get(item) + "\n";
                System.out.print(line);
                litTabFile.write(line);
            }
            litTabFile.close();
        }
        catch (IOException e) {
            e.printStackTrace();
        }
        System.out.println();
    }


    public void printPT() {
        System.out.println("\n--------------------");
        System.out.println("POOL TABLE");
        System.out.println("--------------------");

        try {
            for (int item = 0; item < poolTab.size(); item++) {
                System.out.print(poolTab.get(item));
                poolTabFile.write(poolTab.get(item) + "\n");
            }
            poolTabFile.close();
        }
        catch (IOException e) {
            e.printStackTrace();
        }
        System.out.println();
    }


    public void printIC() {
        System.out.println("\n----------------------------------------------------");
        System.out.println("INTERMEDIATE CODE");
        System.out.println("----------------------------------------------------");
        try {
            for (List<IntermediateCodeEntry> item : IC) {
                StringBuilder line = new StringBuilder();
                for (int i = 0; i < item.size(); i++) {
                    line.append(item.get(i).type).append(item.get(i).value);
                    if (i != item.size() - 1) {
                        line.append("\t");
                    }
                }
                line.append("\n");
                System.out.print(line);
                ICFile.write(line.toString());
            }
            ICFile.close();
        }
        catch (IOException e) {
            e.printStackTrace();
        }
    }

    public static void main(String[] args) throws IOException {
        Pass1 obj = new Pass1();
        obj.compute();
    }
}
